java/src/org/openqa/selenium/support/decorators/WebDriverDecorator.java

```diff
@@ -295,18 +356,36 @@ private Object decorateResult(Object toDecorate) {
     return toDecorate;
   }
 
-  protected final <Z> Z createProxy(final Decorated<Z> decorated, Class<Z> clazz) {
-    Set<Class<?>> decoratedInterfaces = extractInterfaces(decorated);
-    Set<Class<?>> originalInterfaces = extractInterfaces(decorated.getOriginal());
-    Map<Class<?>, InvocationHandler> derivedInterfaces =
-        deriveAdditionalInterfaces(decorated.getOriginal());
+  protected final <Z> Z createProxy(final Decorated<Z> decorated, Class<? extends Z> clazz) {
+    @SuppressWarnings("unchecked")
+    ProxyFactory<Z> factory =
+        (ProxyFactory<Z>)
+            cache.computeIfAbsent(
+                new Definition(decorated), (key) -> createProxyFactory(key, decorated, clazz));
```

Comment:
I am still not quite sure if this is going to work as expected. I believe proxy classes created by ByteBuddy won't be equal, these are different entities in the class loader cache. That is why it probably only make sense to compare between original classes as I assume if OC1 == OC2 then proxy classes created from one of them must also be transitively equal.

Reply:
Yes, they should be transitively equal. This is the basic idea behind this cache, the proxy class can be reused in case the input classes are identical.

The cache is hit, there is a new unit test to ensure this. So i think this is okay, because it does work.

Reply:
that is why it's weird to me as the `decoratedClass == definition.decoratedClass` condition may be actually actually false. Probably, it works because because maps internally use hashCode API, where `return Objects.hash(decoratedClass, originalClass);` obviously computes to true

Reply:
There is indeed a issue with the hashCode implementation. Beside this the Map will call .equals, .hashCode is only to find the bucket where to search.

Reply:
Okay i have replaced the `Definition.hashCode` implementation with a `System.identityHashCode` based implementation .